src/VisualStudio/Core/Def/DocumentOutline/DocumentSymbolViewModel.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    using SymbolKind = LanguageServer.Protocol.SymbolKind;
+
+    internal class DocumentSymbolViewModel : INotifyPropertyChanged
```

Comment:
i don't love that our models are mutable.  That makes reasonaing about things in a multi-threaded environment difficult.  Can we consider moving to immutable data, and instead have a controller that moves between fully consistent immutable snapshots of the data?

Reply:
WPF doesn't work this way. The WPF needs to be able to set data on the ViewModel such as `IsSelected`. If WPF worked with a View-Model-Update pattern this could be done but currently doesn't have this functionality. https://github.com/dotnet/Comet will hopefully make this easy to do when it releases.